servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -103,6 +105,10 @@ data Delayed env c where
              , authD     :: DelayedIO auth
              , bodyD     :: DelayedIO body
              , serverD   :: captures -> auth -> body -> Request -> RouteResult c
+             , cleanupD  :: body -> IO ()
```

Comment:
As I mentioned elsewhere, I don't know exactly why you need the `body` arg. Or why *just* the `body`, for that matter. `DelayedIO (IO ())` seems like a more intuitive signature, to me.